src/controllers/weather_controller.py

```diff
@@ -40,4 +77,98 @@ def get_weather_for_growth_period(self, barley_type, planting_date, lat, lon):
+                current_day_data["max_temperature"] = max_temps[i]
+                current_day_data["min_temperature"] = min_temps[i]
+                current_day_data["precipitation"] = precipitations[i]
+            elif current_date == current_day_data["date"]:
+                current_day_data["max_temperature"] = max(current_day_data["max_temperature"], max_temps[i])
+                current_day_data["min_temperature"] = min(current_day_data["min_temperature"], min_temps[i])
+                current_day_data["precipitation"] += precipitations[i]
+            else:
+                result.append(current_day_data.copy())
+                current_day_data["date"] = current_date
+                current_day_data["max_temperature"] = max_temps[i]
+                current_day_data["min_temperature"] = min_temps[i]
+                current_day_data["precipitation"] = precipitations[i]
+
+        result.append(current_day_data)
+        return result
+
+    @staticmethod
+    def convert_date_format(date_str):
+        if len(date_str) == WeatherController.DATE_LENGTH:
+            return date_str
+        year, month, day = date_str[:4], date_str[4:6], date_str[6:]
+        return f"{year}-{month}-{day}"
+
+    @staticmethod
+    def determine_label(entry):
+        irradiance = entry.get("irradiance", 0)
+        if entry.get("snow") is not None:
+            return "snowy"
+        elif entry.get("precipitation", 0.0) > 0:
+            return "rainy"
+        elif irradiance and float(irradiance) > WeatherController.SUNNY_IRRADIANCE_THRESHOLD:
+            return "sunny"
+        else:
+            return "cloudy"
+
+    def merge_data_based_on_date(self, weather_data, et_data):
+        def convert_to_dict_by_date(data_list):
+            return {self.convert_date_format(data["date"]): data for data in data_list}
+
+
+        def merge_dicts(dict1, dict2):
+            merged = {}
+            for key, value in dict1.items():
+                if key not in merged:
+                    merged[key] = {}
+                merged[key].update(value)
+
+            for key, value in dict2.items():
+                if key not in merged:
+                    merged[key] = {}
+                merged[key].update(value)
+            for key in merged:
+                merged[key]["label"] = WeatherController.determine_label(merged[key])
+
+            return list(merged.values())
+
+        weather_dict = convert_to_dict_by_date(weather_data)
+        et_dict = convert_to_dict_by_date(et_data)
+        return merge_dicts(weather_dict, et_dict)
+
+    def get_merged_weather_data(self, barley_type, start_date, lat, lon):
+        if barley_type == "spring":
+            end_date = (datetime.strptime(start_date, "%Y-%m-%d") + timedelta(days=180)).strftime("%Y-%m-%d")
+        elif barley_type == "winter":
+            end_date = (datetime.strptime(start_date, "%Y-%m-%d") + timedelta(days=270)).strftime("%Y-%m-%d")
```

Comment:
Instead of duplicating the `end_date` declaration twice since the only difference is the "timedelta"

instead, I'd recommend something like this

```
if barley_type == "spring":
   period = 180
elif barley_type == "winter":
   period = 270

end_date = (datetime.strptime(start_date, "%Y-%m-%d") + timedelta(days=period)).strftime("%Y-%m-%d")
```